site/site.go

```diff
@@ -246,28 +271,88 @@ func (h *handler) serveHTML(resp http.ResponseWriter, request *http.Request, req
 	return false
 }
 
-type htmlTemplates struct {
-	tpls *template.Template
-}
-
 // renderWithState will render the file using the given nonce if the file exists
 // as a template. If it does not, it will return an error.
-func (t *htmlTemplates) renderWithState(filePath string, state htmlState) ([]byte, error) {
+func (h *Handler) renderHTMLWithState(rw http.ResponseWriter, r *http.Request, filePath string, state htmlState) ([]byte, error) {
 	var buf bytes.Buffer
 	if filePath == "" {
 		filePath = "index.html"
 	}
-	err := t.tpls.ExecuteTemplate(&buf, filePath, state)
+	tmpl := h.htmlTemplates.Lookup(filePath)
+	if tmpl == nil {
+		return nil, xerrors.Errorf("template %q not found", filePath)
+	}
+
+	// Cookies are sent when requesting HTML, so we can get the user
+	// and pre-populate the state for the frontend to reduce requests.
+	apiKey, actor, _ := httpmw.ExtractAPIKey(rw, r, httpmw.ExtractAPIKeyConfig{
+		Optional: true,
+		DB:       h.opts.Database,
+	})
+	if apiKey != nil && actor != nil {
+		ctx := dbauthz.As(r.Context(), actor.Actor)
+
+		var eg errgroup.Group
+		var user database.User
+		orgIDs := []uuid.UUID{}
+		eg.Go(func() error {
+			var err error
+			user, err = h.opts.Database.GetUserByID(ctx, apiKey.UserID)
+			return err
+		})
+		eg.Go(func() error {
+			memberIDs, err := h.opts.Database.GetOrganizationIDsByMemberIDs(ctx, []uuid.UUID{apiKey.UserID})
+			if errors.Is(err, sql.ErrNoRows) || len(memberIDs) == 0 {
+				return nil
+			}
+			if err != nil {
+				return nil
```

Comment:
If this is on purpose it deserves a comment

Reply:
Fair